### PR TITLE
Canvas: Add universal data link support

### LINKED ISSUE #83871: Canvas: Explore possibility of supporting data links for remainder of canvas elements
Related to https://github.com/grafana/grafana/issues/83346

Currently text model based elements (rectangle, ellipse, metric value, and text) support datalinks.

It probably would be useful though to enable other elements like icon, server, and button to have datalink functionality as well.

### LINKED ISSUE #84143: Canvas: Data links are only supported through field driven text
# Canvas: Data links not working for metric-elements with fields assigned to options other than fields text

What happened?
Data links do not work for metric-elements that have a static text instead of field value even if the field value is associated with background.

It seems that for data links to work appropriately, the first thing is to select a field for text option for metric-value element – this is not expected. After doing so, users can switch text to static one and use field for background color option, and the data links related to the field will work.
In addition, clicking on such metric-value element with field added as an option (other than text, for example as background) and configured data link will break data links for other elements. See minute 2:34 in the screen cast (data links stop working for all elements).
https://github.com/grafana/grafana/assets/141744/bd11c2c0-fb9b-476d-9bdf-1559b725d182

What did you expect to happen?
Expected behaviour is that a metric-value element will display data link for fields that are associated with the element no matter the option (text or background).

Did it work before?
No. This is a new problem discovered after fixing another issue

How do we reproduce it?
1. Add a new element (metric-value).
2. Assing field for background option. Set the text option to static text.
3. Add override with data link for the field.
4. Hovering over such metric-value element will not show data link.
5. Clicking on the element will break previously working data links for any other elements on the canvas already.

Is the bug inside a dashboard panel?
Yes

Environment (with version)?
Grafana: `Grafana v11.0.0-pre ([1352072338](tel:1352072338))`
OS: `macOS 14.3.1 (23D60)`
Browser: `ersion 122.0.6261.94 (Official Build) (arm64)`

### PR DESCRIPTION
**Pull request by nmarrs**
Add universal support for data links in canvas - before data links only worked when field data was tied to text config text for the subset of canvas elements built around the `TextConfig` object. This PR breaks down that limitation, allowing users to create datalinks universally from any element / element property that is tied to data (the exception being the 3 drone elements + the button element which needs probably just a simple on click datalink UX vs the current tooltip approach)

Before

https://github.com/grafana/grafana/assets/22381771/9a09b1ae-04fa-4706-b2b1-239d2469525f

After

https://github.com/grafana/grafana/assets/22381771/febfdc6d-df3d-4da1-8287-f904274f990a

Possible TODOs

- [ ] Clean up data link handling logic (a little verbose at the moment, would be nice to have config specific checks in element code vs in shared util etc)
- [x] Add a gdev dashboard for easier regression testing based on
[this demo dashboard: Data links improvements-1710125144555.json](https://github.com/grafana/grafana/files/14553032/Data.links.improvements-1710125144555.json)

Fixes #83871 
Fixes #84143

### CODE REVIEW

File: public/app/features/canvas/element.ts [hunk 1]
```diff
@@ -72,7 +72,7 @@ export interface CanvasElementItem<TConfig = any, TData = any> extends RegistryI
   /** The default width/height to use when adding  */
   defaultSize?: Placement;
 
-  prepareData?: (ctx: DimensionContext, cfg: TConfig) => TData;
+  prepareData?: (dimensionContext: DimensionContext, elementOptions: CanvasElementOptions<TConfig>) => TData;
```
Comment (nmarrs) [inline public/app/features/canvas/element.ts:75]: This is the primary change / refactor that is allowing the more universal data link support. TLDR; before we were just passing the element's config (specific options for each element) whereas now we are passing down the entire element's options which includes the specific config as well as the the generic container background / border info that can be tied to field data and thus be applicable to data link functionality

File: public/app/plugins/panel/canvas/utils.ts [hunk 2]
```diff
@@ -107,28 +106,155 @@ export function onAddItem(sel: SelectableValue<string>, rootLayer: FrameState |
...
+
+    // Icon config
+    const isIconSVGTiedToFieldData =
+      elementConfig.path?.field &&
+      visibleFields.some((field) => getFieldDisplayName(field, frame) === elementConfig.path?.field);
+    const isIconColorTiedToFieldData =
+      elementConfig.fill?.field &&
+      visibleFields.some((field) => getFieldDisplayName(field, frame) === elementConfig.fill?.field);
+
+    // Wind turbine config (maybe remove / not support this?)
```
Comment (leeoniya) [inline public/app/plugins/panel/canvas/utils.ts:175]: weirdly specific :sweat_smile:

File: public/app/plugins/panel/canvas/utils.ts [hunk 3]
```diff
@@ -107,28 +106,155 @@ export function onAddItem(sel: SelectableValue<string>, rootLayer: FrameState |
...
+      const key = `${link.title}/${link.href}`;
+      if (!linkLookup.has(key)) {
+        links.push(link);
+        linkLookup.add(key);
+      }
+    });
+  }
+};
+
+// TODO: This could be refactored a fair amount, ideally the element specific config code should be owned by each element and not in this shared util file
```
Comment (leeoniya) [inline public/app/plugins/panel/canvas/utils.ts:131]: mhm

was about to suggest DRYing all the copy/paste out, but that would make extracting each to separate elements more work maybe :thinking:

File: public/app/plugins/panel/canvas/utils.ts (hunk 2 above)
Comment (nmarrs) [reply public/app/plugins/panel/canvas/utils.ts:175]: image

Comment (leeoniya) [reply public/app/plugins/panel/canvas/utils.ts:175]: your dad joke game is on point. (get it? on point!?)

**Comment by leeoniya** [comment]
:tada: nice.

top padding of tooltip needs to eat more vegetables.

**Review by baldm0mma — APPROVED** [review]
LGTM! Yeah, I think DRYing out at some point would be nice, but I think getting this in before the deadline is great!

**Pull request merged**